src/protocol/boolean/bit_decomposition.rs

```diff
@@ -0,0 +1,210 @@
+        let a_b = h_b[0..l as usize].to_vec();
+
+        Ok(a_b)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    GenerateRandomBits,
+    RevealAMinusB,
+    AddBtoC,
+    IsPLessThanD,
+    AddDtoG,
+}
+
+impl crate::protocol::Substep for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::GenerateRandomBits => "generate_random_bits",
+            Self::RevealAMinusB => "reveal_a_minus_b",
+            Self::AddBtoC => "add_b_to_c",
+            Self::IsPLessThanD => "is_p_less_than_d",
+            Self::AddDtoG => "add_d_to_g",
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::BitDecomposition;
+    use crate::{
+        error::Error,
+        ff::{Field, Fp31, Fp32BitPrime, Int},
+        protocol::context::SemiHonestContext,
+        protocol::{QueryId, RecordId},
+        test_fixture::{
+            bits_to_value, join3, make_contexts, make_world, share, validate_and_reconstruct,
+            TestWorld,
+        },
+    };
+    use rand::{distributions::Standard, prelude::Distribution, rngs::mock::StepRng, Rng};
+
+    async fn bit_decomposition<F: Field>(
+        ctx: [SemiHonestContext<'_, F>; 3],
+        record_id: RecordId,
+        a: F,
+    ) -> Result<Vec<F>, Error>
+    where
+        Standard: Distribution<F>,
+    {
+        let [c0, c1, c2] = ctx;
+        let mut rand = StepRng::new(1, 1);
+
+        let s = share(a, &mut rand);
+
+        let result = join3(
+            BitDecomposition::execute(c0, record_id, &s[0]),
+            BitDecomposition::execute(c1, record_id, &s[1]),
+            BitDecomposition::execute(c2, record_id, &s[2]),
+        )
+        .await;
+
+        // bit-decomposed values must have the same bit length of the target field
+        assert_eq!(F::Integer::BITS as usize, result[0].len());
+        assert_eq!(F::Integer::BITS as usize, result[1].len());
+        assert_eq!(F::Integer::BITS as usize, result[2].len());
+
+        let bits = (0..result[0].len())
+            .map(|i| validate_and_reconstruct(&result[0][i], &result[1][i], &result[2][i]))
+            .collect::<Vec<_>>();
+
+        Ok(bits)
+    }
+
+    #[ignore]
+    #[tokio::test]
+    pub async fn fp31() -> Result<(), Error> {
+        let world: TestWorld = make_world(QueryId);
+        let ctx = make_contexts::<Fp31>(&world);
+        let [c0, c1, c2] = ctx;
+        let mut rng = rand::thread_rng();
+
+        for i in 0..10 {
+            let input = rng.gen::<Fp31>();
+            let result = bit_decomposition(
+                [c0.clone(), c1.clone(), c2.clone()],
+                RecordId::from(i),
+                input,
+            )
+            .await?;
+
+            if input.as_u128() == 0 {
+                // if the protocol's input is 0, the output could either be
+                // a bitwise sharing of 0 or the prime (in the integers).
+                let x = bits_to_value(&result);
```

Comment:
Actually, I think this is evidence of a bug. I don't think it should return a bit-decomposition of the prime. That's not a valid element of `F_p`. The paper doesn't mention this (as far as I can tell).

Reply:
I also thought it was a bug first, but the paper doesn't mention otherwise either... E.g.,

```
let input a = 0, and random bits in F_p r = 25

c = a - b = -25 mod p = 6
d = c + b = 6 + 25 = 31 <- prime
q = p <? d = 0
f = 225 (2^l - p)
g = f * q = 0
h = d + g = 31
```

This happens because BitwiseSum doesn't take the field prime into account. There are other `r` that outputs the same. i.e., 30, 32, 39. If we don't like this behavior, we could implement a "`F` wrapping" BitwiseSum? I'm not sure how that will affect other protocols though.

Reply:
Wow... I think there's actually an error in the paper. The proof says:

> As for the correctness, notice that `c = a - b mod p` and `d = c + b` (in the integers). Therefore `d = a + qp` for some `q ∈ {0,1}`. Since `a ∈ {0, 1, ..., p-1}` it follows that `q = 1 iff p < d`.

That's incorrect. It should say `q = 1 iff p <= d`.  Consider the case where `q == d`. In that case we have `d = 0 + 1*p`. 

So I think to fix this, we need to compute `p <= d` which is the same thing as `!(d < p)` which we can do with just

```
q = 1 - BIT-LT(d, p)
```